haystack/nodes/ranker/most_diverse.py

```diff
@@ -0,0 +1,146 @@
+        """
+        if query is None or len(query) == 0:
+            raise ValueError("Query is empty")
+        if documents is None or len(documents) == 0:
+            raise ValueError("No documents to choose from")
+
+        diversity_sorted = self.most_diverse_order(query=query, documents=documents)
+        return diversity_sorted[:top_k]
+
+    def most_diverse_order(self, query: str, documents: List[Document]) -> List[Document]:
+        """
+        Orders the given list of documents to maximize diversity. The algorithm first calculates embeddings for
+        each document and the query. It starts by selecting the document that is semantically closest to the query.
+        Then, for each remaining document, it selects the one that, on average, is least similar to the already
+        selected documents. This process continues until all documents are selected, resulting in a list where
+        each subsequent document contributes the most to the overall diversity of the selected set.
+
+        :param query: The search query.
+        :param documents: The list of Document objects to be ranked.
+
+        :return: A list of documents ordered to maximize diversity.
+        """
+
+        # Calculate embeddings
+        doc_embeddings: torch.Tensor = self.model.encode([d.content for d in documents], convert_to_tensor=True)
+        query_embedding: torch.Tensor = self.model.encode([query], convert_to_tensor=True)
+
+        n = len(documents)
+        selected: List[int] = []
+
+        # Compute the similarity matrix between documents
+        doc_sim_matrix: torch.Tensor = doc_embeddings @ doc_embeddings.T
+
+        # Compute the similarity vector between the query and documents
+        query_doc_sim: torch.Tensor = query_embedding @ doc_embeddings.T
+
+        # Start with the document with the highest similarity to the query
+        selected.append(int(torch.argmax(query_doc_sim).item()))
+
+        while len(selected) < n:
+            # Indices of unselected documents
+            unselected: List[int] = [i for i in range(n) if i not in selected]
+
+            # Compute unselected/selected similarity matrix
+            # First iteration through the loop it's shape is (len(unselected), 1)
+            # and then it's (len(unselected), 2) and so on until (1, len(selected)) in the last iteration
+            unselected_selected_sim_matrix = doc_sim_matrix[unselected, :][:, selected]
+
+            # Compute the average similarity score of each unselected document to the selected group of documents
+            avg_sim_unselected: torch.Tensor = torch.mean(unselected_selected_sim_matrix, dim=-1)
+
+            # Select the document with the lowest average similarity score
+            index_unselected = int(torch.argmin(avg_sim_unselected).item())
+            selected.append(unselected[index_unselected])
+
+        ranked_docs: List[Document] = [documents[i] for i in selected]
```

Comment:
I have tested this a bit and couldn't find any bugs. However, it's possible to get of the slicing and make it a bit faster/improve memory usage:

```py
n = len(documents)
selected: List[int] = []

# Compute the similarity vector between the query and documents
query_doc_sim: torch.Tensor = query_embedding @ doc_embeddings.T

# Start with the document with the highest similarity to the query
selected.append(int(torch.argmax(query_doc_sim).item()))

selected_sum = doc_embeddings[selected[0]]

while len(selected) < n:
    # Compute sum of dot products of all selected documents and all other documents
    similarities = selected_sum @ doc_embeddings.T
    # Mask documents that are already selected
    similarities[selected] = torch.inf
    # Select the document with the lowest total similarity score
    index_unselected = int(torch.argmin(similarities).item())
    
    selected.append(index_unselected)
    # It's enough just to add to the selected vectors because dot product is distributive
    selected_sum += doc_embeddings[index_unselected]

ranked_docs: List[Document] = [documents[i] for i in selected]

return ranked_docs
```

This resulted in about a 4x speed improvement with 128 documents and the tests are still passing.